authentication/src/test/java/io/camunda/authentication/filters/OAuth2RefreshTokenFilterTest.java

```diff
@@ -0,0 +1,464 @@
+              } catch (final Exception e) {
+                throw new RuntimeException(e);
+              }
+            });
+
+    thread1.start();
+    thread2.start();
+    thread1.join();
+    thread2.join();
+
+    // Then
+    verify(filterChain, times(2)).doFilter(request, response);
+    // should only call authorize once
+    verify(authorizedClientManager, times(1)).authorize(any(OAuth2AuthorizeRequest.class));
+  }
+
+  @Test
+  void shouldReturnNullForNonOAuth2AuthenticationToken() {
+    // Given
+    final Authentication nonOAuth2Auth =
+        new Authentication() {
+          @Override
+          public Collection<? extends GrantedAuthority> getAuthorities() {
+            return List.of();
+          }
+
+          @Override
+          public Object getCredentials() {
+            return null;
+          }
+
+          @Override
+          public Object getDetails() {
+            return null;
+          }
+
+          @Override
+          public Object getPrincipal() {
+            return null;
+          }
+
+          @Override
+          public boolean isAuthenticated() {
+            return false;
+          }
+
+          @Override
+          public void setAuthenticated(final boolean isAuthenticated)
+              throws IllegalArgumentException {}
+
+          @Override
+          public String getName() {
+            return "";
+          }
+        };
+
+    when(securityContext.getAuthentication()).thenReturn(nonOAuth2Auth);
+
+    // When
+    final OAuth2AuthenticationToken result = filter.getAuthenticationToken();
+
+    // Then
+    assertThat(result).isNull();
+  }
+
+  @Test
+  void shouldDetectExpiredToken() {
+    // Given
+    final OAuth2AuthorizedClient expiredClient = createExpiredAuthorizedClientWithRefreshToken();
+    final OAuth2AuthorizedClient validClient = createValidAuthorizedClient();
+
+    // When & Then
+    assertThat(filter.hasTokenExpired(expiredClient)).isTrue();
+    assertThat(filter.hasTokenExpired(validClient)).isFalse();
+  }
+
+  @Test
+  void shouldDetectRefreshToken() {
+    // Given
+    final OAuth2AuthorizedClient clientWithRefreshToken =
+        createExpiredAuthorizedClientWithRefreshToken();
+    final OAuth2AuthorizedClient clientWithoutRefreshToken =
+        createExpiredAuthorizedClientWithoutRefreshToken();
+
+    // When & Then
+    assertThat(filter.existsRefreshToken(clientWithRefreshToken)).isTrue();
+    assertThat(filter.existsRefreshToken(clientWithoutRefreshToken)).isFalse();
+  }
+
+  @Test
+  void shouldCheckExpirationWithClockSkew() {
+    // Given
+    final Instant expiresAt = Instant.now().plus(30, ChronoUnit.SECONDS);
+    final Instant expiresAtLater = Instant.now().plus(90, ChronoUnit.SECONDS);
+
+    // When & Then
+    assertThat(filter.isExpired(expiresAt)).isTrue();
+    assertThat(filter.isExpired(expiresAtLater)).isFalse();
+  }
+
+  @Test
+  void shouldHandleNullTokenExpiration() {
```

Comment:
Same reasoning as with `shouldReturnNullForNonOAuth2AuthenticationToken`, I suggest to test this case by invoking `doFilterInternal` instead.